#**Maestría en Inteligencia Artificial Aplicada**
##**Curso: Inteligencia Artificial y Aprendizaje Automático**
###Tecnológico de Monterrey
###Prof Luis Eduardo Falcón Morales

### **Semana 9**
####**Actividad en clase: Taxonomía de métricas de clasificación**

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv("oil_spill.csv", header=None)
df.head()

0      1        2       3    4           5      6     7        8     9   \
0   1   2558  1506.09  456.63   90   6395000.0  40.88  7.89  29780.0  0.19   
1   2  22325    79.11  841.03  180  55812500.0  51.11  1.21  61900.0  0.02   
2   3    115  1449.85  608.43   88    287500.0  40.42  7.34   3340.0  0.18   
3   4   1201  1562.53  295.65   66   3002500.0  42.40  7.97  18030.0  0.19   
4   5    312   950.27  440.86   37    780000.0  41.43  7.03   3350.0  0.17   

   ...       40        41       42       43     44  45        46     47    48  \
0  ...  2850.00   1000.00   763.16   135.46   3.73   0  33243.19  65.74  7.95   
1  ...  5750.00  11500.00  9593.48  1648.80   0.60   0  51572.04  65.73  6.26   
2  ...  1400.00    250.00   150.00    45.13   9.33   1  31692.84  65.81  7.84   
3  ...  6041.52    761.58   453.21   144.97  13.33   1  37696.21  65.67  8.07   
4  ...  1320.04    710.63   512.54   109.16   2.58   0  29038.17  65.66  7.35   

   49  
0   1  
1   0  
2   1  
3   1  
4   0  

[5 rows x 50 columns]

In [ ]:
X, y = df.iloc[:, :-1], df.iloc[:, -1]

print(X.shape)
print(y.shape)

(937, 49)
(937,)


In [ ]:
print((X.shape[0] - y.sum()) / y.shape[0])
print(y.sum() / y.shape[0])

0.9562433297758804
0.04375667022411953


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 937 entries, 0 to 936
Data columns (total 50 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       937 non-null    int64  
 1   1       937 non-null    int64  
 2   2       937 non-null    float64
 3   3       937 non-null    float64
 4   4       937 non-null    int64  
 5   5       937 non-null    float64
 6   6       937 non-null    float64
 7   7       937 non-null    float64
 8   8       937 non-null    float64
 9   9       937 non-null    float64
 10  10      937 non-null    float64
 11  11      937 non-null    float64
 12  12      937 non-null    float64
 13  13      937 non-null    float64
 14  14      937 non-null    float64
 15  15      937 non-null    float64
 16  16      937 non-null    float64
 17  17      937 non-null    float64
 18  18      937 non-null    float64
 19  19      937 non-null    float64
 20  20      937 non-null    float64
 21  21      937 non-null    float64
 22  22

#**Hecho 1-a: Describe el problema y el objetivo que se busca**

*   Esta base de datos fue parte del artículo de 1998 llamado "Machine Learning for the Detection of Oil Spills in Satellite Radar Images", Kubat, Holte, Matwin:

https://link.springer.com/content/pdf/10.1023/A:1007452223027.pdf

*   La manera en que se generó la base de datos fue a partir de imágenes satelitales del océano, unas de las cuales incluían derrame de petróleo y otras no lo incluían. Cada imagen se seccionó y a cada pedazo se le aplicaron técnicas de procesamiento de imágenes para extraer diversos vectores de características que describen el contenido de ese pedazo de imagen.

*   Se tienen 47 columnas de vectores de características como datos de entrada, las etiquetas de 1s y 0s en el vector y. Hay un total de 937 registros.

*   Todos los vectores de características son de tipo numérico real (flotante). No existen datos perdidos.

*   El problema se plantea como de aprendizaje supervisado, donde el objetivo es determinar si existe o no derrame a partir de los vectores de características numéricos de entrada que se obtuvieron de cada pedazo de imagen.

* El predecir que hay un derrame, implica movilizar un flotilla de barcos con equipo especial, técnicos e ingenieros especialistas, helicópteros, etc.)



#**HECHO 1-b: ¿Es un problema desbalanceado? ¿Cómo se podría enfrentar?**

*   Existe un desbalanceo de clases: 95.6% (clase negativa) vs 4.4% (clase positiva).

*   Observamos que tenemos un conjunto de datos bastante desbalanceado. 

*   El desbalanceo es tal que se requiere algo más que simplemente técnicas de ajuste de pesos de clases a través del argumento **weight_class** que tienen algunas modelos de aprendizaje automático.

*   Esto nos lleva a que deberemos usar técnicas de balanceo como SMOTE, Tomeklinks y no solo weight_class. 

*   El desbalanceo también no lleva a seleccionar algunas métricas que se centren en la clase positiva.



In [ ]:
y[0:7]

0    1
1    0
2    1
3    1
4    0
5    1
6    1
Name: 49, dtype: int64

In [ ]:
print(X.shape, y.shape)

print(y.sum())

print((y.shape[0] -y.sum()) / y.shape[0])

print(y.sum() / y.shape[0])


(937, 49) (937,)
41
0.9562433297758804
0.04375667022411953


#**Hecho 1-c:  Implicaciones a que conllevan un FP o un FN**

**Falso Positivo:** Una predicción que resulte en un Falso Positivo significa hacer una movilización de recursos muy costosos: movilización de helicópteros, barcos, maquinaria para ello, técnicos e ingenieros especialistas, etc. Lo cual implicararía un gran gasto no justificado.

**Falso Negativo:** Una predicción que resulte en un Falso Negativo, implicaría tiempo muy valioso para detener el derrame de petróleo, con un gran impacto tanto ecológico como económico (mayor cantidad de petróleo derramado que habrá que limpiar).

*   Es decir, en este caso podemos concluir que ambos tipos de errores llevarían a costos de diferente naturaleza muy elevados, pero "igualmente" costosos.



#**Hecho 1-d: Determinar el tipo de métrica de acuerdo a la Taxonomía de los autores**

Es decir, determinar si la salida probabilística es clave para el problema o simplemente la predicción sobre las clases.

*   Así, debemos cuidar ambos tipos de errores, pero como las clases están desbalanceada, habrá que hacerlo con métricas como el recall y la especificiada, y su promedio geométrico G-mean.  

*   En este caso el valor probabilístico en sí no sería de importancia ya que ambos errores nos llevan a un mismo error igualmente costoso. Es decir, lo primordial es determinar si hubo o no derrame, por lo que la salida se centra en un problema de predecir clases sin importar mucho su valor probabilístico.

*   En resumen, bajo estos criterios, nos decidiríamos por una métrica de umbral (threshold metric), en particular la métrica G-mean, la cual toma en consideración que ambos errores FN y FP son igualmente costosos.

###**Fin de la Actividad en clase de la semana 9**